### PR TITLE
Code missing final newline was highlighted incorrectly.

### PR DESCRIPTION
**Pull request by k3rni**
Addresses #3050.

The issue here was that, for input missing the final newline, the final increment line step was not performed. This time we always finish the line on end of token stream, possibly introducing a blank one, but then strip it out entirely. If the final line was not blank, the strip does nothing.

The swift example file included contains @codesman's submission code verbatim. If using it here due to copyrights is an issue, we can easily replace it with something else. Also note that it has no final newline (and should not have one), which is important to this issue.

### CODE REVIEW

File: test/fixtures/code_formatting/swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
+//
+//  PhoneNumber.swift
+//  PhoneNumber
+//
+//  Created by Tom Holland on 8/13/16.
+//  Copyright © 2016 Tom Holland. All rights reserved.
```
Comment (Insti) [inline test/fixtures/code_formatting/swift]: Taking code other peoples (non open source) code is a big no-no.
In any case it would be better to recreate the issue using a the minimal possible example code.

File: test/exercism/html_formatter_test.rb [hunk 2]
```diff
@@ -28,6 +29,16 @@ def test_python_docstring
...
     assert_equal '', @doc.css('span#L5').text.rstrip, "Line 5 should be blank"
+
+  end
+
+  def test_swift
+    load_sample 'swift', 'swift'
+
+    assert_equal 121, @doc.css('span[id^=L]').size, "Formatting should preserve linecount"
+    assert_equal "}", @doc.css('span#L121').text.rstrip, "Line 121 should have only a closing brace"
+    assert_equal "    }", @doc.css('span#L120').text.rstrip, "Line 120 should have a four space indent and closing brace"
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: Don't use an instance variable `@doc` here (or anywhere,) when you can use a locally scoped return value from a method call. `doc = load_sample ...`

File: test/exercism/html_formatter_test.rb [hunk 3]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
   end
 
   def load_sample(name, language)
-    code = load_example name
-    @raw_lines = code.lines
+    @code = load_example name
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `@code` should not be an instance variable.

File: test/exercism/html_formatter_test.rb [hunk 4]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
...
+    @raw_lines = @code.lines
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `@raw_lines` should not be an instance variable.

File: test/exercism/html_formatter_test.rb [hunk 5]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
...
     lexer = get_lexer language
 
-    @lexemes = lexer.lex(code)
+    @lexemes = lexer.lex(@code)
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `@lexemes` should not be an instance variable.

File: test/exercism/html_formatter_test.rb [hunk 6]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
...
     @output = @formatter.format(@lexemes)
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `@output` should not be an instance variable.

File: test/exercism/html_formatter_test.rb [hunk 7]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
...
 
     @doc = Loofah::HTML::DocumentFragment.parse(@output)
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `@doc` should not be an instance variable.

Comment (k3rni) [inline test/exercism/html_formatter_test.rb]: All of these aid in debugging. I don't have strong opinions on these, but test code is not public API, and such constructs are acceptable IMO. Will change.

File: test/exercism/html_formatter_test.rb [hunk 8]
```diff
@@ -37,11 +48,11 @@ def get_lexer(language)
   end
 
   def load_sample(name, language)
```
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: `load_sample` should be returning the tokens that you will be using as input to the thing you are testing.
In this case the `lexemes`. The actual call to the thing you are testing should be happening within the test.

File: test/exercism/html_formatter_test.rb (hunk 7 above)
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: I do have strong opinions on these :)

You can do whatever you like while you're developing, but when you check it in test code should be held to equal standard of code quality as the rest of the code. (This includes running linters like Rubocop on it, and paying attention to code complexity issues highlighted by tools like `flog`.)

File: test/exercism/html_formatter_test.rb (hunk 2 above)
Comment (Insti) [inline test/exercism/html_formatter_test.rb]: Don't hide your call to the thing you're testing in a method.

`doc = @formatter.format(lexemes)`

Should be here. 

(I'm personally not a huge fan of setting up the thing you're testing in `setup`, but understand why some people choose to do it, so am not objecting to the `@formatter` instance variable in this case.)

**Comment by Insti** [comment]
This PR does not revert the changes made to these test files:

`test/fixtures/approvals/markdown_double_braces.approved.txt`
`test/exercism/converts_markdown_to_html_test.rb`
 other
I believe the original versions were correct and the current versions are not.

File: lib/rouge/formatters/html_exercism.rb [hunk 9]
```diff
@@ -128,12 +128,21 @@ def formatted_code(tokens, line_counter)
             formatted << end_line << start_line(line_counter)
           end
         end
-        formatted.gsub(start_line(line_counter), '')
+        adjust_final_lines formatted, line_counter
+      end
+
+      def adjust_final_lines(formatted, line_counter)
```
Comment (Insti) [inline lib/rouge/formatters/html_exercism.rb]: This is just a patch over some fundamental parsing issues and will add extra lines to solutions in some cases.
Github won't let me comment on bits of code that have not changed here, so I might need to go back to the original PR and add annotations there.

**Comment by Insti** [comment]
@k3rni sorry if I'm appearing too negative here.

I'm really grateful that you've chosen to work on this issue. 

What you have done mostly works, but some of the choices you have made are making it difficult to create the tests that can identify the exact problems that various 'blank line' edge cases are causing.

My hope is that if we get the tests into a useful state, we can then create smaller more focused test cases to ensure that the formatter is doing the right thing in all cases.

This will also require changes to the formatter, but good focused tests should guide us to an improved solution.

:heart:

File: lib/rouge/formatters/html_exercism.rb (hunk 9 above)
Comment (Insti) [inline lib/rouge/formatters/html_exercism.rb]: (This is one of the reasons I'm in favour of revert and re-apply rather than patch-forwards.)

**Comment by k3rni** [comment]
One of these edgecases that currently fails is, as you pointed out:
`test_markdown_code_with_double_braces`. However, I believe the current state is correct, and the previous one was wrong, let me explain:

input to Markdown parser: ```three-leftquotes \n {{x: y}} \n three-leftquotes``` (it's hard to paste three-leftquotes correctly)
input to `formatted_code` from lexer: 
```
[[<Token Punctuation>, "{"],
 [<Token Error>, "{x:"],
 [<Token Text.Whitespace>, " "],
 [<Token Error>, "y"],
 [<Token Punctuation>, "}"],
 [<Token Error>, "}"],
 [<Token Text.Whitespace>, "\n"]]
```

There is only one newline here, and it's a terminating one. Therefore, this code has only one line, and not two as was specified previously.

**Comment by k3rni** [comment]
However, all of this may be a blind alley. Rouge advertises:

```
Rouge::Formatters::HTMLLinewise.new(formatter, class_format: 'line-%i') This formatter will split your code into lines, each contained in its own div. The class_format option will be used to add a class name to the div, given the line number.
```

I think we should consider using this instead. I may make a new PR.

**Comment by Insti** [comment]
+:100: on using one of Rouge's built-in formatters instead of our own custom one.

**Comment by Insti** [comment]
> `test_markdown_code_with_double_braces`. However, I believe the current state is correct,

Thanks for the explanation, I now agree with you that it is correct and it is only 1 line.

**Comment by Insti** [comment]
> Rouge::Formatters::HTMLLinewise
> I think we should consider using this instead. I may make a new PR.

Can you at least try it out and report back, I'm interested in knowing if it will work.
Let me know if you need any help.

**Comment by k3rni** [comment]
@Insti take a look here: https://github.com/k3rni/exercism.io/blob/linewise-formatter/lib/rouge/formatters/html_exercism.rb

This is a straight-up rewrite using `token_lines` from `Rouge::Formatters::HTMLLinewise`, because the class by itself is not flexible enough. This still fails some tests - same ones my original solution failed initially.

* `test_markdown_code_with_double_braces` - but as we discussed above, the test might be wrong
* `test_markdown_with_clojure_code` - a similar issue with whitespace now being wrapped differently, notably this time it expects the trailing newlines to be wrapped but not leading spaces.

If this is a good direction, I'll turn it into a PR.

**Comment by Insti** [comment]
This looks much simpler than the previous version, nice work.

> because the class (Rouge::Formatters::HTMLLinewise) by itself is not flexible enough

Can you say a little about why HTMLLinewise is not good enough?

**Comment by k3rni** [comment]
* no option to change element used, always `<div>`
* no option to use `id` instead of `class`

**Comment by Insti** [comment]
The Exercism version has a LOT of options, how many of those are actually used by the app?
The more of them we can eliminate the simpler things become.

**Comment by k3rni** [comment]
The only place Exercism uses this formatter directly is in [syntax_highlighter.rb:42](/exercism/exercism.io/blob/master/lib/exercism/syntax_highlighter.rb#L42). That in turn is used by Markdown renderer for code blocks and directly for a `syntax` ERB helper. None of these customize anything but the language. Thus:

1. the `line_numbers` option is always used, we can throw out `stream_untableized`
2. and `line_numbers_id` is never changed from default
3. `inline_theme` is never used
4. line numbering always starts at 1

I suggest removing code for the first three.

**Comment by Insti** [comment]
>  I suggest removing code for the first three.
Great! That should simplify things a bit more.

**Comment by Insti** [comment]
The thing to keep in mind here is we're not trying to build a generic formatter, we're trying to build an Exercism specific formatter, so it only needs to do the things that Exercism needs it to do, not everything that it could possibly do.

**Comment by kytrinyx** [comment]
+100 on simplifying and removing anything that we don't need.

**Comment by k3rni** [comment]
Obsoleted by #3063.
